compiler/rustc_trait_selection/src/traits/select/mod.rs

```diff
@@ -473,7 +473,15 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
             }
         } else {
             let has_non_region_infer = stack.obligation.predicate.has_non_region_infer();
-            if let Some(candidate) = self.winnow_candidates(has_non_region_infer, candidates) {
+            let did = stack.obligation.predicate.def_id();
+            let is_sizedness_or_auto_or_default_predicate = self.tcx().is_sizedness_trait(did)
+                || self.tcx().trait_is_auto(did)
+                || self.tcx().is_default_trait(did);
```

Comment:
I dislike this as is :thinking: maybe have a shared used by both old and new solver (with comments)
```rust
enum CandidatePreferenceMode {
    Default,
    Marker,
}

impl CandidatePreferenceMode {
    fn compute<I: Interner>(cx: I, trait_def_id: I::DefId) -> CandidatePreferenceMode;
}
```